### PR TITLE
Add types to please mypy

### PR DESCRIPTION
**Pull request by henri-gasc**
While I am at this, I am tempted to run black & isort (and fix the line ending) to have consistency across the codebase

### CODE REVIEW

**Comment by einarf** [comment]
Definitely something we need. I already did basic black formatting + config and started adding more annotations in some places.

**Comment by henri-gasc** [comment]
From what I can see, we do not need `from __future__ import annotations` as `Dict`, `Tuple`, etc. are deprecated since [3.9](https://docs.python.org/3/library/typing.html#typing.Dict).
We also have to make sure to use `Union[type1, type2]` and not `type1 | type2` as this syntax is available for 3.10+.

I also corrected what I think are mistakes, so I want you to look at them closely (I will comment on them when I finish with the `loader/*` files).

**Comment by einarf** [comment]
The future import includes [Postponed evaluation of annotations (That is default in 3.10)](https://peps.python.org/pep-0563/).

Will not result in `NameError: name 'A' is not defined`

```py
>>> from __future__ import annotations
>>> class A:
...     def test() -> A:
...         pass
```

Support for for `|` operator for types (without import: `TypeError: unsupported operand type(s) for |: 'type' and 'type'`)

```py
>>> from __future__ import annotations
>>> class A: pass
...
>>> class B: pass
...
>>> def test() -> A | B:
...     pass
```

Can replace `Optional` 

```py
>>> from __future__ import annotations
>>> def test() -> float | None:
...     pass
```

The `|` operator doesn't work for types is enclosed in `""` or `''`. in cases were `typing.TYPE_CHECKING` is used. That needs python 3.10+. However, the world will not end if `Union` is still used even if it's unnecessary in most cases.

**Comment by henri-gasc** [comment]
Okay I did not know they were backported

**Comment by einarf** [comment]
Definitely correct mistakes if you see them. The initial typing of the project was in 3.7 and it's lacking with today's standards 😄

If any imperfections we can fix that in 3.x.x.

File: moderngl_window/atlas/base.py [hunk 1]
```diff
@@ -44,21 +45,21 @@ class AtlasImage(BaseImage):
     """An atlas image using Pillow"""
 
     def __init__(self, image: Image):
-        self.image = image
+        self._image = image
```
Comment (henri-gasc) [inline moderngl_window/atlas/base.py:48]: I think the underscore was forgotten here

File: moderngl_window/atlas/base.py [hunk 2]
```diff
@@ -44,21 +45,21 @@ class AtlasImage(BaseImage):
...
 
     @property
     def width(self) -> int:
-        return self._image.size[0]
+        return self._image.width
```
Comment (henri-gasc) [inline moderngl_window/atlas/base.py:52]: PIL.Image.Image has `width` and `height` proterty, so it is a mistakes to not use them

File: moderngl_window/atlas/simple_atlas.py [hunk 3]
```diff
@@ -116,13 +115,13 @@ def __init__(
...
 
         # The physical size limit for the current hardware
-        self._max_size = self._ctx.info["GL_MAX_VIEWPORT_DIMS"]
+        self._max_size: tuple[int, int] = self._ctx.info["GL_MAX_VIEWPORT_DIMS"]
 
         # Atlas content
         self._texture = self._ctx.texture(self.size, components=self._components)
 
         # We want to be able to render into the atlas texture
-        self._fbo = self._fbo = self._ctx.framebuffer(color_attachments=[self._texture])
```
Comment (henri-gasc) [inline moderngl_window/atlas/simple_atlas.py:125]: There is no reason there should be `self._fbo = self._fbo = ...` apart from a mistake

File: moderngl_window/capture/base.py [hunk 4]
```diff
@@ -21,28 +22,28 @@ class BaseVideoCapture:
 
     def __init__(
         self,
-        source: Union[moderngl.Texture, moderngl.Framebuffer] = None,
```
Comment (henri-gasc) [inline moderngl_window/capture/base.py:24]: A source has to be defined, so treat it as such

**Comment by henri-gasc** [comment]
Some general comments:
I tried to change most of the `if var:` and `if not var:` to the correct form (so `if var is not None`, `if var != ""`, etc.), but I probably forgot some of those.
I did not touch the documentation (except to change the types)
I did not go back and use `future` to replace the `Optional` and `Union` with `|` and `| None`

File: moderngl_window/atlas/base.py (hunk 1 above)
Comment (einarf) [reply moderngl_window/atlas/base.py:48]: Yup. It's sadly incomplete code I need to clean up soon

File: moderngl_window/atlas/base.py (hunk 2 above)
Comment (einarf) [reply moderngl_window/atlas/base.py:52]: It's probably better for readability.

File: moderngl_window/atlas/simple_atlas.py (hunk 3 above)
Comment (einarf) [reply moderngl_window/atlas/simple_atlas.py:125]: yup :)

File: moderngl_window/capture/base.py (hunk 4 above)
Comment (einarf) [reply moderngl_window/capture/base.py:24]: Definitely a good idea

**Comment by einarf** [comment]
Thanks a lot for this pull request. There are a couple of odd things such as `Generator` not having 3 arguments but this is not critical and won't affect anything runtime. For the sake of progress I will just merge this right now so I can rebase by local branches for more fixes.

**Comment by einarf** [comment]
Did another black + ruff pass and modernized the workflow. It should also run (again) on PRs. master branch passes basic code linting now at least and at least doc tests 👍 (Will need to get unit tests running at some point)

**Pull request merged**